src/main/scala/org/clulab/habitus/utils/ContextExtractor.scala

```diff
@@ -0,0 +1,260 @@
+    val mostFreqDate0Sent = extractContext(doc, allEventMentions, 0, "DATE", entityHistogram)
+    val mostFreqDate1Sent = extractContext(doc, allEventMentions, 1, "DATE", entityHistogram)
+    val mostFreqDateOverall = extractContext(doc, allEventMentions, Int.MaxValue, "DATE", entityHistogram)
+    val mostFreqCrop0Sent = extractContext(doc, allEventMentions, 0, "CROP", entityHistogram)
+    val mostFreqCrop1Sent = extractContext(doc, allEventMentions, 1, "CROP", entityHistogram)
+    val mostFreqCropOverall = extractContext(doc, allEventMentions, Int.MaxValue, "CROP", entityHistogram)
+    val mostFreqFertilizer0Sent = extractContext(doc, allEventMentions, 0, "FERTILIZER", entityHistogram)
+    val mostFreqFertilizer1Sent = extractContext(doc, allEventMentions, 1, "FERTILIZER", entityHistogram)
+    val mostFreqFertilizerOverall = extractContext(doc, allEventMentions, Int.MaxValue, "FERTILIZER", entityHistogram)
+
+    //for each event mention, get the sentence id, and map it to a case class called contextDetails, which will have all of mostFreq* information
+    createSentidContext(sentidContext, mostFreqLocation0Sent, mostFreqLocation1Sent, mostFreqLocationOverall,
+      mostFreqDate0Sent, mostFreqDate1Sent, mostFreqDateOverall, mostFreqCrop0Sent, mostFreqCrop1Sent,
+      mostFreqCropOverall,mostFreqFertilizer0Sent, mostFreqFertilizer1Sent,
+      mostFreqFertilizerOverall)
+
+    sentidContext
+  }
+
+  def findMostFreqContextEntitiesForAllEvents(mentionContextMap: mutable.Map[Mention, Seq[EntityDistFreq]], howManySentAway:Int, entityType:String):Seq[MostFreqEntity] = {
+    mentionContextMap.keys.toSeq.map(key=>findMostFreqContextEntitiesForOneEvent(key,mentionContextMap(key), entityType,howManySentAway))
+  }
+
+  //if key exists add+1 to its value, else add 1 as its value
+  def checkAddFreq(map: mutable.Map[String, Int], key: String, freq: Int): Int = {
+    val newFreq = map.getOrElse(key, 0) + freq
+
+    map(key) = newFreq
+    newFreq
+  }
+
+  //given a user input (e.g.,LOC,1-- which means find which Location occurs most frequently within 1 sentence of this
+  // event), calculate it from available frequency and sentence data
+  def findMostFreqContextEntitiesForOneEvent(mention:Mention, contexts:Seq[EntityDistFreq], entityType:String, howManySentAway:Int):MostFreqEntity= {
+    val entityFreq = mutable.Map[String, Int]()
+    var maxFreq = 0
+    var mostFreqEntity = ""
+    //go through each of the contexts and find if any of them satisfies the condition in the query
+    for (ctxt <- contexts) {
+      for (sentFreq <- ctxt.entityDistFrequencies) {
+        val sentDist = sentFreq._1
+        val freq = sentFreq._2
+        if (sentDist <= howManySentAway && ctxt.nerTag.contains(entityType)) {
+          val newFreq = checkAddFreq(entityFreq, ctxt.entityValue, freq)
+          if (newFreq >= maxFreq) {
+            maxFreq = newFreq
+            mostFreqEntity = ctxt.entityValue
```

Comment:
An alternative, not necessarily better, is
```scala
val mostFrequentEntity = if (entityFreq.nonEmpty) Some(entityFreq.maxBy(_._2)._1) else None
```
There are fewer special values like 0 and "".  Later Scalas have maxByOption to take care of the empty condition.